csrc/python_frontend/translation.cpp

```diff
@@ -0,0 +1,524 @@
+    NVF_ERROR(op->isA<LoadStoreOp>() || is_cast_op);
+
+    size_t input_fd_index = map_val_to_fd_index_.at(op->input(0));
+
+    // DataType::Index does not exist in python_frontend, so convert to
+    // DataType::Int
+    DataType scalar_dtype = op->output(0)->dtype();
+    if (scalar_dtype == DataType::Index) {
+      scalar_dtype = DataType::Int;
+    }
+
+    if (op->input(0)->isA<TensorView>()) {
+      Tensor output =
+          fd_->defineTensor(op->output(0)->as<TensorView>()->nDims());
+      map_val_to_fd_index_.emplace(op->output(0), output());
+      fd_->defineRecord(new CastOpRecord<TensorView*, TensorView*>(
+          {fd_->recordingState(input_fd_index)},
+          {fd_->recordingState(output())},
+          "ops.cast",
+          serde::RecordType::CastTv,
+          static_cast<TensorView* (*)(DataType, TensorView*)>(castOp),
+          std::get<PrimDataType>(scalar_dtype.type)));
+    } else {
+      Scalar output = fd_->defineScalar();
+      map_val_to_fd_index_.emplace(op->output(0), output());
+      fd_->defineRecord(new CastOpRecord<Val*, Val*>(
+          {fd_->recordingState(input_fd_index)},
+          {fd_->recordingState(output())},
+          "ops.cast",
+          serde::RecordType::CastVal,
+          static_cast<Val* (*)(DataType, Val*)>(castOp),
+          std::get<PrimDataType>(scalar_dtype.type)));
+    }
+  }
+
+  // Map BinaryOp to python_frontend OpRecord
+  void handle(const BinaryOp* bop) final {
+    bool is_lhs_tv = bop->lhs()->isA<TensorView>();
+    bool is_rhs_tv = bop->rhs()->isA<TensorView>();
+
+    if (is_lhs_tv || is_rhs_tv) {
+      Tensor output = fd_->defineTensor(bop->out()->as<TensorView>()->nDims());
+      map_val_to_fd_index_.emplace(bop->out(), output());
+
+      if (is_lhs_tv && is_rhs_tv) {
+        handleOpRecord<nvfuser::BinaryOp>(
+            bop,
+            serde::RecordType::Binary_TV,
+            bop->out()->as<TensorView>(),
+            bop->lhs()->as<TensorView>(),
+            bop->rhs()->as<TensorView>());
+      } else if (is_lhs_tv && !is_rhs_tv) {
+        handleOpRecord<nvfuser::BinaryOp>(
+            bop,
+            serde::RecordType::Binary_TV_VAL,
+            bop->out()->as<TensorView>(),
+            bop->lhs()->as<TensorView>(),
+            bop->rhs());
+      } else {
+        handleOpRecord<nvfuser::BinaryOp>(
+            bop,
+            serde::RecordType::Binary_VAL_TV,
+            bop->out()->as<TensorView>(),
+            bop->lhs(),
+            bop->rhs()->as<TensorView>());
+      }
+    } else {
+      Scalar output = fd_->defineScalar();
+      map_val_to_fd_index_.emplace(bop->out(), output());
+      handleOpRecord<nvfuser::BinaryOp>(
+          bop,
+          serde::RecordType::Binary_VAL,
+          bop->out(),
+          bop->lhs(),
+          bop->rhs());
+    }
+  }
+
+  // Add Broadcast operation to FusionDefinition
+  void handle(const BroadcastOp* bcast_op) final {
```

Comment:
I think we might actually need to do some pattern matching here since `broadcast_in_dim` translates to broadcast+expand. You could always just match the `ExpandOp` individually but then if you actually input a `broadcast_in_dim` from python and did a round trip you would wind up with _two_ `broadcast_in_dim` calls I think.

Reply:
Future PRs will add the `void handle(const ExpandOp* eop)` for expand operations. To keep the visitor pattern clean, it would be easier to add an `ExpandOpRecord` and `expand` to the python-frontend.

Reply:
> To keep the visitor pattern clean, it would be easier to add an `ExpandOpRecord` and `expand` to the python-frontend.

I agree.